[reland] Add default Saved Variable hooks

Varal7 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #62564
* __->__ #62563

Summary:

Expose a pair of functions to Python users: torch.autograd.graph.set_saved_tensors_default_hooks(pack, unpack) and torch.autograd.graph.reset_saved_tensors_default_hooks().
These functions control the hooks applied to saved tensors: all tensors saved in that context will be packed using the pack function, then unpacked accordingly when needed.

Currently, this works by simply calling register_hooks (cf #60975) directly at the end of the constructor of a SavedVariable. This could be optimized further by not performing the copy before registering default hooks, but this would require a small refactor. Edit: the refactor is done in #61927.

A current limitation is that if users create tensors in this context, they will not be able to register additional hooks on the saved tensor.

For instance, to perform something like #28997, one could define a pack function that saves to disk whenever the tensor size is too big and returns a filename, then unpack simply reads the content of the file and outputs a tensor, e.g.:

```
def pack(x):
    name = os.path.join(tmp_dir, str(uuid.uuid4()))
    torch.save(x, name)
    return name

def unpack(name):
    return torch.load(name)
```

Relanding previous PR: https://github.com/pytorch/pytorch/pull/61834

Original PR led to timeout error in: https://www.internalfb.com/mast/job/yuguo-release_canary_offline_training-inlinecvrp_a-canary_offline_train_28a7ecfc

PR reverting: https://github.com/pytorch/pytorch/pull/62415

Now passing: https://www.internalfb.com/mast/job/quach-release_canary_offline_training-inlinecvrp_a-canary_offline_train_9bb57e98

The difference with the new version is that when calling`PyDefaultSavedVariableHooks::get_hooks`, we only acquire the GIL if the hooks are defined.

Differential Revision: [D30045405](https://our.internmc.facebook.com/intern/diff/D30045405)

Varal7 committed:
[reland] Add default Saved Variable hooks

Summary:

Expose a pair of functions to Python users: torch.autograd.graph.set_saved_tensors_default_hooks(pack, unpack) and torch.autograd.graph.reset_saved_tensors_default_hooks().
These functions control the hooks applied to saved tensors: all tensors saved in that context will be packed using the pack function, then unpacked accordingly when needed.

Currently, this works by simply calling register_hooks (cf #60975) directly at the end of the constructor of a SavedVariable. This could be optimized further by not performing the copy before registering default hooks, but this would require a small refactor. Edit: the refactor is done in #61927.

A current limitation is that if users create tensors in this context, they will not be able to register additional hooks on the saved tensor.

For instance, to perform something like #28997, one could define a pack function that saves to disk whenever the tensor size is too big and returns a filename, then unpack simply reads the content of the file and outputs a tensor, e.g.:

```
def pack(x):
    name = os.path.join(tmp_dir, str(uuid.uuid4()))
    torch.save(x, name)
    return name

def unpack(name):
    return torch.load(name)
```

Relanding previous PR: https://github.com/pytorch/pytorch/pull/61834

Original PR led to timeout error in: https://www.internalfb.com/mast/job/yuguo-release_canary_offline_training-inlinecvrp_a-canary_offline_train_28a7ecfc

Now passing: https://www.internalfb.com/mast/job/quach-release_canary_offline_training-inlinecvrp_a-canary_offline_train_9bb57e98

The difference with the new version is we don't need to acquire the GIL when calling `PyDefaultSavedVariableHooks::get_hooks`.

[ghstack-poisoned]

Varal7 commented on torch/csrc/autograd/python_saved_variable_hooks.cpp:
```diff
@@ -44,4 +44,37 @@ namespace torch { namespace autograd {
+  }
+
+  std::unique_ptr<SavedVariableHooks> PyDefaultSavedVariableHooks::get_hooks() {
+    if (!pack_hook_ || !unpack_hook_) {
```
This does not need to acquire the GIL

Varal7 commented on torch/csrc/autograd/python_saved_variable_hooks.cpp:
```diff
@@ -44,4 +44,37 @@ namespace torch { namespace autograd {
+    if (!pack_hook_ || !unpack_hook_) {
+      return nullptr;
+    }
+    py::function pack_hook = py::reinterpret_borrow<py::function>(pack_hook_);
```
Here, we need the GIL

albanD commented on torch/csrc/autograd/python_saved_variable_hooks.cpp:
```diff
@@ -44,4 +44,37 @@ namespace torch { namespace autograd {
+    }
+    py::function pack_hook = py::reinterpret_borrow<py::function>(pack_hook_);
+    py::function unpack_hook = py::reinterpret_borrow<py::function>(unpack_hook_);
+    return std::make_unique<PySavedVariableHooks>(pack_hook, unpack_hook);
```
Note that this constructor actually steals the refererence from these two functions. But if you borrow above, then you actually stole the reference from the fields directly.
I feel like there is a missing incref here.

albanD replied:
These namings :D
borrow actually incref and steal does not.

But isn't the `py::function` a managed type that will get decref automatically? Should you use `pack_hook.release()`?

Varal7 replied:
Actually, the `py::function` gets `Py_XDECREF` when it goes out of scope (https://github.com/pybind/pybind11/blob/6d4409466b6c7e6fa53e7b93cb05d78647c26313/include/pybind11/pytypes.h#L199) so we're fine.

If we wanted to use `pack_hook.release()` here, we could do so by adding a new constructor `PySavedVariableHooks:: PySavedVariableHooks(PyObject* pack_hook, PyObject* unpack_hook)` which takes `py::handle` / `PyObject*` (they're the same) as argument.

albanD replied:
Ho right these arguments are py::function so there is an extra incref when you call the functions...
Mixing pybind and PyObject is tricky!

Varal7 committed:
Update on "[reland] Add default Saved Variable hooks"


Summary:

Expose a pair of functions to Python users: torch.autograd.graph.set_saved_tensors_default_hooks(pack, unpack) and torch.autograd.graph.reset_saved_tensors_default_hooks().
These functions control the hooks applied to saved tensors: all tensors saved in that context will be packed using the pack function, then unpacked accordingly when needed.

Currently, this works by simply calling register_hooks (cf #60975) directly at the end of the constructor of a SavedVariable. This could be optimized further by not performing the copy before registering default hooks, but this would require a small refactor. Edit: the refactor is done in #61927.

A current limitation is that if users create tensors in this context, they will not be able to register additional hooks on the saved tensor.

For instance, to perform something like #28997, one could define a pack function that saves to disk whenever the tensor size is too big and returns a filename, then unpack simply reads the content of the file and outputs a tensor, e.g.:

```
def pack(x):
    name = os.path.join(tmp_dir, str(uuid.uuid4()))
    torch.save(x, name)
    return name

def unpack(name):
    return torch.load(name)
```

Relanding previous PR: https://github.com/pytorch/pytorch/pull/61834

Original PR led to timeout error in: https://www.internalfb.com/mast/job/yuguo-release_canary_offline_training-inlinecvrp_a-canary_offline_train_28a7ecfc

Now passing: https://www.internalfb.com/mast/job/quach-release_canary_offline_training-inlinecvrp_a-canary_offline_train_9bb57e98

The difference with the new version is we don't need to acquire the GIL when calling `PyDefaultSavedVariableHooks::get_hooks`.

[ghstack-poisoned]

albanD reviewed (approved):
LGTM

Varal7 commented:
@varal7 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30045405).
